src/mixnet.rs

```diff
@@ -0,0 +1,211 @@
+use anyhow::{anyhow, Error};
+use async_channel::{self, Receiver, Sender};
+use futures::{future::Future, FutureExt, SinkExt, StreamExt};
+use nym_sphinx::addressing::clients::Recipient;
+use nym_websocket::{requests::ClientRequest, responses::ServerResponse};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio::net::TcpStream;
+use tokio_tungstenite::{
+    connect_async, tungstenite::protocol::Message, MaybeTlsStream, WebSocketStream,
+};
+use tracing::debug;
+
+use crate::message::*;
+
+/// Mixnet implements a read/write connection to a Nym websockets endpoint.
+pub(crate) struct Mixnet {
+    // the websocket connection between us and the Nym endpoint we're using.
+    ws_stream: WebSocketStream<MaybeTlsStream<TcpStream>>,
+
+    // a channel of inbound messages from the endpoint.
+    // the transport reads from (listens) to the receiver of this channel.
+    inbound_tx: Sender<InboundMessage>,
+
+    // a channel of outbound messages to be written to the endpoint.
+    // the transport writes to the sender of this channel.
+    outbound_rx: Receiver<OutboundMessage>,
+}
+
+impl Mixnet {
+    pub(crate) async fn new(
+        uri: &String,
+    ) -> Result<(Mixnet, Receiver<InboundMessage>, Sender<OutboundMessage>), Error> {
+        let (ws_stream, _) = connect_async(uri).await.map_err(|e| anyhow!(e))?;
+        let (inbound_tx, inbound_rx): (Sender<InboundMessage>, Receiver<InboundMessage>) =
+            async_channel::unbounded();
```

Comment:
Tokio has its own channel implementation
https://docs.rs/tokio/latest/tokio/sync/mpsc/fn.unbounded_channel.html

I tried to removed async_channel's and uses the tokio's provided channel, but there are just a lot of errors to deal with.

The `async_channel` crate seems to be coming from `smol` runtime. Mixing different runtime libraries in rust has never been a smooth experience for me.

Reply:
update to use tokio channels, helped me remove some of the tokio tasks, thanks!!